src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_SIZE = 10;
+    private T[] elements;
+    private int size;
 
+    public ArrayList() {
+        elements = (T[]) (new Object[DEFAULT_SIZE]);
+    }
+
+    public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size >= elements.length) {
+            elements = getExpanded();
+        }
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("You can't add an element at that index");
+        } else if (index == size) {
+            elements[index] = value;
+        } else {
+            elements = insertElement(value, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("There is no element with such index");
+        } else {
+            return elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (get(index) != null) {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldValue = get(index);
+        elements = removeAndCopy(index);
+        size--;
+        return oldValue;
```

Comment:
use System.arraycopy to move all the elements 1 position to the left

hint - you can transfer elements from and to the same array